src/main/java/net/minecraftforge/common/capabilities/CapabilityProvider.java

```diff
@@ -5,206 +5,158 @@
-    private @Nullable CapabilityDispatcher capabilities;
-    private boolean valid = true;
-
-    private boolean                       isLazy             = false;
-    private Supplier<ICapabilityProvider> lazyParentSupplier = null;
-    private CompoundTag                   lazyData           = null;
-    private boolean initialized = false;
-
-    protected CapabilityProvider(Class<B> baseClass)
-    {
-        this(baseClass, false);
-    }
+    @Nullable
+    private CapabilityDispatcher<T> attachedCaps = null;
+    protected boolean capsValid = true;
+    protected CompoundTag lazyCapNbt = null;
+    protected boolean capsInitialized = false;
 
-    protected CapabilityProvider(final Class<B> baseClass, final boolean isLazy)
+    @SuppressWarnings("unchecked")
+    private void doGatherCapabilities()
     {
-        this.baseClass = baseClass;
-        this.isLazy = SUPPORTS_LAZY_CAPABILITIES && isLazy;
+        this.capsInitialized = true;
+        this.attachedCaps = ForgeEventFactory.gatherCapabilities((AttachCapabilitiesEvent<T>) this.getCapEvent());
     }
 
-    protected final void gatherCapabilities()
-    {
-        gatherCapabilities(() -> null);
-    }
-
-    protected final void gatherCapabilities(@Nullable ICapabilityProvider parent)
-    {
-        gatherCapabilities(() -> parent);
-    }
-
-    protected final void gatherCapabilities(@Nullable Supplier<ICapabilityProvider> parent)
+    /**
+     * This method is called from {@link ItemStack#copy()} and {@link ServerPlayer#restoreFrom(ServerPlayer, boolean)} to clone 
+     * the capabilities of the original.
+     * <p>
+     * <b>It is not designed for use elsewhere, and may break or crash if used improperly.</b>
+     */
+    @SuppressWarnings("unchecked")
+    protected <P extends CapabilityProvider<T>> void copyCapsFrom(P other)
     {
-        if (isLazy && !initialized)
+        if(other.capsInitialized)
         {
-            lazyParentSupplier = parent == null ? () -> null : parent;
-            return;
+            this.capsInitialized = true;
+            this.attachedCaps = other.getDispatcher().copy((T) this);
         }
-
-        doGatherCapabilities(parent == null ? null : parent.get());
+        else this.lazyCapNbt = other.lazyCapNbt;
     }
 
-    private void doGatherCapabilities(@Nullable ICapabilityProvider parent)
-    {
-        this.capabilities = ForgeEventFactory.gatherCapabilities(baseClass, getProvider(), parent);
-        this.initialized = true;
-    }
+    protected abstract AttachCapabilitiesEvent<? extends T> getCapEvent();
 
-    @NotNull
-    B getProvider()
-    {
-        return (B)this;
-    }
-
-    protected final @Nullable CapabilityDispatcher getCapabilities()
+    /**
+     * <b>This method will initialize capabilities when called.</b>
+     * @return The {@link CapabilityDispatcher}, or null, if nothing attached.
+     */
+    protected final @Nullable CapabilityDispatcher<T> getDispatcher()
```

Comment:
I'd recommend naming this `getCapDispatcher`, just for consistency with the rest of the methods.
```suggestion
    protected final @Nullable CapabilityDispatcher<T> getCapDispatcher()
```

Reply:
Yep, good point.  I wasn't using that name because I was using it for a different function for the same signature, but that method was removed, freeing the name up.